packages/docs-infra/src/useUrlHashState/useUrlHashState.ts

```diff
@@ -0,0 +1,162 @@
+'use client';
+import * as React from 'react';
+
+export interface UseUrlHashStateOptions {
+  /**
+   * Whether to automatically read the hash on mount
+   * @default true
+   */
+  readOnMount?: boolean;
+
+  /**
+   * Whether to automatically watch for hash changes
+   * @default true
+   */
+  watchChanges?: boolean;
+
+  /**
+   * Custom hash parser function
+   * @default (hash) => hash.slice(1) // Remove the '#'
+   */
+  parseHash?: (hash: string) => string;
+
+  /**
+   * Custom hash formatter function
+   * @default (value) => value
+   */
+  formatHash?: (value: string) => string;
+}
+
+export interface UseUrlHashStateResult {
+  /**
+   * Current hash value (without the '#')
+   */
+  hash: string | null;
+
+  /**
+   * Set a new hash value (will update the URL)
+   * @param value - New hash value
+   * @param replace - Whether to use replaceState instead of pushState
+   */
+  setHash: (value: string | null, replace?: boolean) => void;
+
+  /**
+   * Whether the initial hash has been processed
+   */
+  hasProcessedInitialHash: boolean;
+
+  /**
+   * Whether the user has explicitly interacted with the hash
+   */
+  hasUserInteraction: boolean;
+
+  /**
+   * Mark that user has interacted with the hash
+   */
+  markUserInteraction: () => void;
+}
+
+const defaultParseHash = (hash: string): string => hash.slice(1); // Remove the '#'
+const defaultFormatHash = (value: string): string => value;
+
+/**
+ * Hook for managing URL hash state with SSR support
+ */
+export function useUrlHashState(options: UseUrlHashStateOptions = {}): UseUrlHashStateResult {
```

Comment:
Let's keep these publicly exported hooks simple and to the point. The options are unnecessary, and same for anything else than `hash` and `setHash`. What's the point of `hasUserInteraction`? It's doing nothing beyond wrapping `React.useState`